### PR TITLE
Ahoyapps 957 recording start stop

### PR DESCRIPTION
**Pull request by timmydoza**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [AHOYAPPS-957](https://issues.corp.twilio.com/browse/AHOYAPPS-957)

### Description

This PR adds the start/stop recording functionality to the app. See it live here: https://start-stop-recording-video-app-6813-5075-dev.twil.io

(Slack me for the passcode)

In order to use this branch with our quick deploy plugin, you'll need to use [this](https://github.com/twilio-labs/plugin-rtc/pull/55) branch of the plugin as well. 

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [ ] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with `npm test`
* [ ] Manually sanity tested running locally
* [ ] Included screenshot as PR comment (if needed)
* [ ] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Re-tested if necessary

### CODE REVIEW

**Comment by charliesantos** [comment]
UX wise, I think stopping a recording should be a one step process and not have to click More > Stop recording.
I just find myself looking for the stop button the first time I used it :). Thoughts?

File: src/components/MenuBar/Menu/Menu.tsx [hunk 1]
```diff
@@ -42,12 +54,21 @@ export default function Menu(props: { buttonClassName?: string }) {
           horizontal: 'center',
         }}
       >
+        <RecordingButton onClick={() => setMenuOpen(false)} setIsRecordingSnackbarOpen={setIsRecordingSnackbarOpen} />
+
+        <MenuItem onClick={() => setSettingsOpen(true)}>
+          <IconContainer>
+            <SettingsIcon />
+          </IconContainer>
+          <Typography variant="body1">Conference Settings</Typography>
```
Comment (charliesantos) [inline src/components/MenuBar/Menu/Menu.tsx:63]: The button is labeled "Conference Settings", but the dialog box is "Audio and Video Settings". Should we make this consistent?

File: src/state/usePasscodeAuth/usePasscodeAuth.ts [hunk 2]
```diff
@@ -69,6 +70,32 @@ export default function usePasscodeAuth() {
     [user]
   );
 
+  const updateRecordingRules = useCallback(
```
Comment (charliesantos) [inline src/state/usePasscodeAuth/usePasscodeAuth.ts:73]: I remember we talked about this duplicate code. Can you please remind me what we agreed on?

**Review by charliesantos — CHANGES_REQUESTED** [review]

**Comment by charliesantos** [comment]
It works @timmydoza !
Few things that might need more discussion @aaalaniz @timmydoza 

- If stop recording is not clicked, recording is still on going after all participants have left. Will this automatically stop at some point?
- Any participant can start/stop recording. Is there a concept of "admin" who can start/stop the recording?
- Per this [document](https://docs.google.com/document/d/1MMbX3Y7VgP_kajMJXpRLdBvb-s3XoBiXCvc6EbgB5oE/edit), `A user should see a UI element before joining a room that their media may be recorded`. I don't see any indicator/UI element prior joining.
- If all participants are already in the room, and when a user starts recording, we only show the recording indicator and proceed with recording anyway. Should we present a UI element to make this more apparent and allow the other participants whether they want to get recorded or not?

**Review by mhuynh5757 — APPROVED** [review]

**Comment by timmydoza** [comment]
> * If stop recording is not clicked, recording is still on going after all participants have left. Will this automatically stop at some point?

Good question! I assume that the recording will stop when the room ends, which would be 5 minutes after all participants leave. I can double check this. 

> * Any participant can start/stop recording. Is there a concept of "admin" who can start/stop the recording?

No, we have decided not to worry about roles here. Any user should be able to start/stop the recording. 

> * Per this [document](https://docs.google.com/document/d/1MMbX3Y7VgP_kajMJXpRLdBvb-s3XoBiXCvc6EbgB5oE/edit), `A user should see a UI element before joining a room that their media may be recorded`. I don't see any indicator/UI element prior joining.
> * If all participants are already in the room, and when a user starts recording, we only show the recording indicator and proceed with recording anyway. Should we present a UI element to make this more apparent and allow the other participants whether they want to get recorded or not?

I didn't see anything like this in the [Abstract designs](https://app.abstract.com/share/1a1c4fbe-23d5-4891-9a05-6b610f748cf3) that Travis made. We can check in with @aaalaniz or Travis to figure out what we should do here with respect to a pre-join indicator.

**Comment by aaalaniz** [comment]
Great feedback @charliesantos! 

This application is a demo app so we are intentionally not designing user roles for the start/stop recording actions. However, you provide a good point about informing the user about the recording indicator.

@timmydoza what do you think about adding a tooltip to the recording indicator and the start/stop recording button with the following message:

> All participant's audio and video tracks are recorded by default. This application's recording controls are only intended for testing and development purposes.

The copy might need some work. I'll loop in the appropriate people to get feedback.

File: src/components/MenuBar/Menu/Menu.tsx (hunk 1 above)
Comment (timmydoza) [reply src/components/MenuBar/Menu/Menu.tsx:63]: We could - I'll check with Travis.

File: src/state/usePasscodeAuth/usePasscodeAuth.ts (hunk 2 above)
Comment (timmydoza) [reply src/state/usePasscodeAuth/usePasscodeAuth.ts:73]: Sure! Once we move from the `video-app-service` over to twilio functions (as part of [AHOYAPPS-964](https://issues.corp.twilio.com/browse/AHOYAPPS-964), which I will start this sprint), I'll be able to update `src/state/useFirebaseAuth` so that it consumes the new API. Since `usePasscodeAuth` and `useFirebaseAuth` will be consuming the same token server (with different authentication mechanisms), I'll be able to refactor those two hooks so that they share some logic. Additionally, I'd like to bring in the axios library, which will help us deal with errors a little better. 

Additionally a custom `useAxios` hook (or similar) will help to cut back on some of this duplicated code. Something like this:
`const {isRequesting, data, error} = useAxios('/token')`

In a nutshell, I will refactor much of `src/state/` when we move off video-app-service.

**Comment by timmydoza** [comment]
> Great feedback @charliesantos!
> 
> This application is a demo app so we are intentionally not designing user roles for the start/stop recording actions. However, you provide a good point about informing the user about the recording indicator.
> 
> @timmydoza what do you think about adding a tooltip to the recording indicator and the start/stop recording button with the following message:
> 
> > All participant's audio and video tracks are recorded by default. This application's recording controls are only intended for testing and development purposes.
> 
> The copy might need some work. I'll loop in the appropriate people to get feedback.

We can add a tooltip. I can check in with Travis to get his thoughts. I'm not sure that "All participant's audio and video tracks are recorded by default." is best though. It makes it sound like all tracks are recorded all the time. 

Also, do you want to try and handle this use case?

>A user should see a UI element before joining a room that their media may be recorded

File: src/state/usePasscodeAuth/usePasscodeAuth.ts (hunk 2 above)
Comment (charliesantos) [reply src/state/usePasscodeAuth/usePasscodeAuth.ts:73]: Got it. Can you please add a small TODO above this line so I don't forget next time I look at this part of the code? 😄

**Comment by timmydoza** [comment]
> UX wise, I think stopping a recording should be a one step process and not have to click More > Stop recording.
> I just find myself looking for the stop button the first time I used it :). Thoughts?

I agree. I can check in with Travis to get his thoughts. Maybe there's a reason why he designed it the way he did. 

Since the recording controls are for 'testing and development purposes' (as Aaron puts it), I don't think it's important that they are completely frictionless. I don't think they will be commonly used - so I personally think that they are ok as is.

File: src/components/MenuBar/Menu/Menu.tsx (hunk 1 above)
Comment (timmydoza) [reply src/components/MenuBar/Menu/Menu.tsx]: I reverted this to 'Audio and Video Settings' per Travis.

File: src/state/usePasscodeAuth/usePasscodeAuth.ts (hunk 2 above)
Comment (timmydoza) [reply src/state/usePasscodeAuth/usePasscodeAuth.ts:73]: Well, I plan on refactoring more than this one function. I'll refactor a lot of stuff in `src/state`. Plus, it looks like this will be the next ticket I pick up. If I add a TODO it won't be there for long 🙂

Comment (charliesantos) [reply src/state/usePasscodeAuth/usePasscodeAuth.ts:73]: Ok, nvm then :)

**Comment by timmydoza** [comment]
@charliesantos This is ready for a review again. Per the Slack discussion, I added the following. 

1. A confirmation dialog when the Start Recording button is clicked:

image

2. A tooltip on the recording indicator:

image


These features have been deployed to a new URL. I'll send it to you on Slack.

File: src/components/MenuBar/Menu/Menu.tsx (hunk 1 above)
Comment (charliesantos) [reply src/components/MenuBar/Menu/Menu.tsx]: Screen Shot 2021-01-12 at 10 34 08 AM
I see the code has been updated, but the deployed app is showing this ☝️ 
Do you have the latest deployed?

**Review by charliesantos — APPROVED** [review]
